From: Beniamino Galvani <bgalvani@redhat.com>
Date: Wed, 14 Feb 2018 11:43:00 +0100
Subject: settings: preserve agent-owned secrets on connection update
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64

QWZ0ZXIgd3JpdGluZyB0aGUgY29ubmVjdGlvbiB0byBkaXNrIGFuZCByZXJlYWRpbmcgaXQsIGlu
IGFkZGl0aW9uIHRvCnJlc3RvcmluZyBhZ2VudC1vd25lZCBzZWNyZXRzIGluIHRoZSBjYWNoZSB3
ZSBtdXN0IGFsc28gcmVzdG9yZQphZ2VudC1vd25lZCBzZWNyZXRzIGZyb20gdGhlIG9yaWdpbmFs
IGNvbm5lY3Rpb25zIHNpbmNlIHRoZXkgYXJlIGxvc3QKZHVyaW5nIHRoZSB3cml0ZS4KClJlcG9y
dGVkLWJ5OiBNw6RydCBCYWtob2ZmIDxhbm9uQHNpZ2lsLnJlZD4KCmh0dHBzOi8vYnVnemlsbGEu
Z25vbWUub3JnL3Nob3dfYnVnLmNnaT9pZD03OTMzMjQKKGNoZXJyeSBwaWNrZWQgZnJvbSBjb21t
aXQgZjljNTBiZjNkM2UxZTUyZDU4MDNkNTVmYTk3YmY1NjkzMGZkMzAyMCkK
---
 src/settings/nm-settings-connection.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/settings/nm-settings-connection.c b/src/settings/nm-settings-connection.c
index 37a0b3a..5862637 100644
--- a/src/settings/nm-settings-connection.c
+++ b/src/settings/nm-settings-connection.c
@@ -608,6 +608,7 @@ nm_settings_connection_update (NMSettingsConnection *self,
 	gboolean replaced = FALSE;
 	gs_free char *logmsg_change = NULL;
 	GError *local = NULL;
+	gs_unref_variant GVariant *con_agent_secrets = NULL;
 
 	g_return_val_if_fail (NM_IS_SETTINGS_CONNECTION (self), FALSE);
 
@@ -658,9 +659,20 @@ nm_settings_connection_update (NMSettingsConnection *self,
 	    && !nm_connection_compare (NM_CONNECTION (self),
 	                               replace_connection,
 	                               NM_SETTING_COMPARE_FLAG_EXACT)) {
+		gs_unref_object NMConnection *simple = NULL;
+
 		if (log_diff_name)
 			nm_utils_log_connection_diff (replace_connection, NM_CONNECTION (self), LOGL_DEBUG, LOGD_CORE, log_diff_name, "++ ");
 
+		/* Make a copy of agent-owned secrets because they won't be present in
+		 * the connection returned by plugins, as plugins return only what was
+		 * reread from the file. */
+		simple = nm_simple_connection_new_clone (NM_CONNECTION (self));
+		nm_connection_clear_secrets_with_flags (simple,
+		                                        secrets_filter_cb,
+		                                        GUINT_TO_POINTER (NM_SETTING_SECRET_FLAG_AGENT_OWNED));
+		con_agent_secrets = nm_connection_to_dbus (simple, NM_CONNECTION_SERIALIZE_ONLY_SECRETS);
+
 		nm_connection_replace_settings_from_connection (NM_CONNECTION (self), replace_connection);
 
 		replaced = TRUE;
@@ -688,6 +700,8 @@ nm_settings_connection_update (NMSettingsConnection *self,
 				g_variant_unref (dict);
 			}
 		}
+		if (con_agent_secrets)
+			(void) nm_connection_update_secrets (NM_CONNECTION (self), NULL, con_agent_secrets, NULL);
 	}
 
 	nm_settings_connection_recheck_visibility (self);
